kolibri/core/content/signals.py

```diff
@@ -22,3 +23,15 @@ def reorder_channels_upon_deletion(sender, instance=None, *args, **kwargs):
     For a given channel, decrement the order of all channels that come after this channel.
     """
     ChannelMetadata.objects.filter(order__gt=instance.order).update(order=F('order') - 1)
+
+
+@receiver(pre_delete, sender=ChannelMetadata)
+def update_lesson_resources_before_delete(sender, instance=None, *args, **kwargs):
+    # Update the resources array of all lessons to ensure they don't have
+    # any deleted content
+    lessons = Lesson.objects.all()
```

Comment:
Could possibly reduce the search space for lessons by doing a pre-filter of the lessons here:
`lessons = Lesson.objects.filter(resources__contains=instance.id)`
So you will now only iterate over lessons that contain this channel_id in the resources JSON blob.